File: ruby/leet_code/2441.rb
# Module: LeetCode
# The module encapsulates solutions for problems from LeetCode.
module LeetCode
  # Function: find_max_k
  # The function finds the maximum positive integer in the array that also has its negative in the array.
  #
  # @param nums [Array<Integer>] The array of integers to search through.
  #
  # @return [Integer] The maximum positive integer that also has its negative in the array, or -1 if no such integer exists.
  #
  # @example
  #   find_max_k([3, -1, 1, -3, 2]) #=> 3
  #
  # This function does not handle errors as it expects the input to always be an array of integers.
  def find_max_k(nums)
    # Initialize the maximum positive integer to -1
    max_positive = -1
    # Initialize an empty hash to keep track of visited numbers
    visited = {}

    # Iterate over each number in the array
    nums.each do |num|
      # If the number is positive, its negative has been visited, and it's greater than the current maximum positive
      if num > 0 && visited.key?(-num) && num > max_positive
        # Update the maximum positive
        max_positive = num
        # If the number is negative, its positive has been visited, and its absolute value is greater than the current maximum positive
      elsif num < 0 && visited.key?(-num) && -num > max_positive
        # Update the maximum positive
        max_positive = -num
      end
      # Mark the number as visited
      visited[num] = true
    end

    # Return the maximum positive integer
    max_positive
  end
end

File: rust/src/leet_code/2441.rs
//! # LeetCode Problem 2441
//!
//! This module contains the solution for LeetCode problem 2441. The main function in this module is `find_max_k`,
//! which finds the maximum integer `k` such that both `-k` and `k` exist in the given array.
//!
//! # Examples
//!
//! ```
//! let nums = vec![3, 2, -2, 5, -3];
//! assert_eq!(Solution::find_max_k(nums), 3);
//! ```

impl Solution {
    /// Finds the maximum integer `k` such that both `-k` and `k` exist in the given array.
    ///
    /// This function iterates over the given array and keeps track of the positive and negative numbers it has seen.
    /// If it encounters a number `n` and `-n` has already been seen (or vice versa), it updates the maximum `k`.
    ///
    /// # Parameters
    ///
    /// * `nums: Vec<i32>` - The input array of 32-bit integers.
    ///
    /// # Returns
    ///
    /// * `i32` - Returns the maximum `k` such that both `-k` and `k` exist in the array. If no such `k` exists, it returns -1.
    ///
    /// # Errors
    ///
    /// This function does not return any errors. If the input array is empty, it simply returns -1.
    pub fn find_max_k(nums: Vec<i32>) -> i32 {
        // Boolean arrays to keep track of the positive and negative numbers we have seen
        let mut seen_neg = vec![false; 1001];
        let mut seen_pos = vec![false; 1001];

        // The maximum `k` we have found so far
        let mut max = -1;

        // Iterate over the numbers in the array
        for n in nums {
            if n > 0 {
                // If the number is positive, mark it as seen in the positive array
                seen_pos[n as usize] = true;
                // If the negative counterpart has been seen, update the maximum `k`
                if seen_neg[n as usize] {
                    max = max.max(n);
                }
            } else {
                // If the number is negative, mark it as seen in the negative array
                seen_neg[-n as usize] = true;
                // If the positive counterpart has been seen, update the maximum `k`
                if seen_pos[-n as usize] {
                    max = max.max(-n);
                }
            };
        }

        // Return the maximum `k` we have found
        max
    }
}
